torch/utils/checkpoint.py

```diff
@@ -354,7 +354,11 @@ def inner_unpack(packed):
                     #  detached_inputs = detach_variable(tuple(inputs))
                     with torch.enable_grad(), torch.cuda.amp.autocast(self.had_autocast_in_fwd):
                         with torch.autograd.graph.saved_tensors_hooks(inner_pack, inner_unpack):
+                            if isinstance(self.function, torch.nn.Module):
+                                self.function._recomputing = True
```

Comment:
For DDP to work properly with this implementation of checkpointing, it seems that DDP needs to know whether the forward pass is being recomputed or computed for the first time. 

For the current checkpointing support with reentrant autograd, this is a non-issue because some core counting logic in DDP's forward pass is only enabled if `torch.is_grad_enabled()` is True, and one of the fwd passes run by reentrant autograd checkpointing is run under torch.no_grad().

However this implementation does not use torch.no_grad(), hence the checks in DDP's forward pass no longer work. Adding this and checking it wherever `torch.is_grad_enabled()` (as seen in the changes in this PR) makes it work, but not sure if this is a great design.

Any thoughts? @zhaojuanmao @albanD

Reply:
This looks a bit hacky to me.
If we thing that users will really need to know this, then we should allow them to give a context manager or hooks that we will call around the re-computation here.

Reply:
Sure, I think a context manager supplied by the user could be a good solution.

I do think that we need this as the current checkpointing implementation works for DDP due to running with no_grad, new checkpoint will have issues if there is no way to know about the recomputing.

Reply:
Yes, we can add this feature to the existing implementation as well (even though DDP doesn't need it).

Reply:
Although, not sure if this is ideal from DDP user perspective. The user would do something like

model = DDP(...)

# forward pass
checkpoint(model, inputs)

and if they needed to pass a context manager into checkpoint(), it would leak some of DDP's internal abstractions. And since checkpoint is called after DDP is applied, there doesn't seem to be a way to have DDP "tell" checkpoint that it needs to be able to distinguish between recomputing or not.